va/validation-authority.go

```diff
@@ -169,9 +223,20 @@ func (va ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdentif
 		// We don't expect to make multiple requests to a client, so close
 		// connection immediately.
 		DisableKeepAlives: true,
+		// Intercept Dial in order to connect to the IP address we
+		// select.
+		Dial: dialer,
 	}
+
 	logRedirect := func(req *http.Request, via []*http.Request) error {
-		va.log.Info(fmt.Sprintf("validateSimpleHTTP [%s] redirect from %q to %q", identifier, via[len(via)-1].URL.String(), req.URL.String()))
+		challenge.Redirects = append(challenge.Redirects, req.URL.String())
+		dialer, addrUsed, err := va.resolveAndConstructDialer(req.URL.Host, req.URL.Scheme)
+		challenge.ResolvedAddrs = append(challenge.ResolvedAddrs, addrUsed)
```

Comment:
Combining the addresses for all redirects into one big bag is confusing. I would propose the structure look like this:

```
{
 "type": "simpleHttp",
 ...
 "validationData": {
  "simpleHttp": [{
    "url": "http://foo.example.com/.well-known/acme-certificate/abc.txt",
    "addressesFound": [ "123.45.67.9" ],
    "addressesUsed": [ "123.45.67.9" ]
  }, {
   "url": "http://other.example.com/acme-validator/cba.txt",
    "addressesFound": [ "33.44.55.66", "33.44.55.67" ],
    "addressesUsed": [ "33.44.55.66" ]
  }]
 }
}
```

Each challenge type would potentially have a different set of validation data.